<a href="https://colab.research.google.com/github/ik-okoro/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_t_tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [1]:
# Import relevant libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy import stats

In [2]:
# Load dataset
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data

--2020-09-07 20:19:11--  https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 25936 (25K) [application/x-httpd-php]
Saving to: ‘imports-85.data’

imports-85.data     100%[===================>]  25.33K  --.-KB/s    in 0.07s   

2020-09-07 20:19:11 (390 KB/s) - ‘imports-85.data’ saved [25936/25936]



In [6]:
column_headers = ['symboling', 'normalized-losses', 'make', 'fuel-type', 
                'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 
                'engine-location', 'wheel-base', 'length', 'width', 'height', 
                'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 
                'fuel-system', 'bore', 'stroke', 'compression-ratio', 
                'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']

df_auto = pd.read_csv("imports-85.data", names=column_headers)

print(df_auto.shape)

df_auto.head()

(205, 26)


symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [7]:
# Reflect missing values with NaN
df_auto.replace({"?" : np.NaN})

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0            3               NaN  alfa-romero  ...       21          27  13495
1            3               NaN  alfa-romero  ...       21          27  16500
2            1               NaN  alfa-romero  ...       19          26  16500
3            2               164         audi  ...       24          30  13950
4            2               164         audi  ...       18          22  17450
..         ...               ...          ...  ...      ...         ...    ...
200         -1                95        volvo  ...       23          28  16845
201         -1                95        volvo  ...       19          25  19045
202         -1                95        volvo  ...       18          23  21485
203         -1                95        volvo  ...       26          27  22470
204         -1                95        volvo  ...       19          25  22625

[205 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [8]:
# Find population mean
df_auto["highway-mpg"].mean()

30.75121951219512

In [9]:
# Create sample size with 10 cars
df_auto_sample = df_auto.sample(10, random_state=30)

df_auto_sample.head()

symboling normalized-losses           make  ... city-mpg highway-mpg  price
121          1               154       plymouth  ...       31          38   6692
176         -1                65         toyota  ...       27          32  10898
24           1               148          dodge  ...       31          38   6229
88          -1               137     mitsubishi  ...       23          30   9279
72           3               142  mercedes-benz  ...       16          18  35056

[5 rows x 26 columns]

### Null Hypothesis:

The sample mean is not different from the population mean. $H_0: \overline{x} = \overline\mu$

### Alternative Hypothesis:
The sample mean is different from the population mean. $H_1: \overline{x} \neq \overline\mu$

### Confidence Level: 95%

In [10]:
stats.ttest_1samp(df_auto_sample["highway-mpg"], df_auto["highway-mpg"].mean())

Ttest_1sampResult(statistic=0.8912422391619923, pvalue=0.39600040771093326)

### Conclusion:
With a t-statistic of 0.89 and a p-value of 0.396, I fail to reject the null hypothesis that the sample mean is different from the population mean.

I cannot say they are different.

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

### Null Hypothesis:

The cars typically have a fuel efficiency of 35 mpg on the highway. $H_0: \overline{x} = \text{35 mpg}$

### Alternative Hypothesis:
The cars do not typically have a fuel efficiency of 35 mpg on the highway. $H_1: \overline{x} \neq \text{35 mpg}$

### Confidence Level: 95%

In [11]:
# Calculate t-score

stats.ttest_1samp(df_auto_sample["highway-mpg"], 35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

### Conclusion:

With a t-score of -0.957 and a p value of 0.364, I fail to reject the null hypothesis that the cars typically have a fuel efficiency of 35 mpg on the highway with 95% certainty.

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [13]:
# Bigger sample size (n=100)

df_auto_sample = df_auto.sample(100, random_state=30)

df_auto_sample.head()

symboling normalized-losses           make  ... city-mpg highway-mpg  price
121          1               154       plymouth  ...       31          38   6692
176         -1                65         toyota  ...       27          32  10898
24           1               148          dodge  ...       31          38   6229
88          -1               137     mitsubishi  ...       23          30   9279
72           3               142  mercedes-benz  ...       16          18  35056

[5 rows x 26 columns]

### Null Hypothesis:

The cars typically have a fuel efficiency of 35 mpg on the highway. $H_0: \overline{x} = \text{35 mpg}$

### Alternative Hypothesis:
The cars do not typically have a fuel efficiency of 35 mpg on the highway. $H_1: \overline{x} \neq \text{35 mpg}$

### Confidence Level: 95%

In [14]:
# Calculate t-score

stats.ttest_1samp(df_auto_sample["highway-mpg"], 35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

### Conclusion:

With a t-score of greater than 4 (-6.52), and a p-value of 3e-09, I reject the null hypothesis that the cars typically have a fuel efficiency of 35 mpg on the highway with 95% certainty.

I suggest they typically have a fuel efficiency less than 35 mpg on the highway.

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

Larger sample sizes give a better reflection of the population and help give more accurate predictions

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [15]:
# Load the data
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-09-07 22:26:55--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.06s   

2020-09-07 22:26:55 (286 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [17]:
# Load and alter the dataframe

column_headers = ["party", "handicapped-infants", "water-project", "budget", 
                 "physician-fee-freeze", "el-salvador-aid", "religious-groups", 
                 "anti-satellite-ban", "aid-to-contras", "mx-missile", "immigration", 
                 "synfuels", "education", "right-to-sue", "crime", "duty-free", "south-africa"]

df = pd.read_csv("house-votes-84.data", names=column_headers)

df = df.replace({"y" : 1, "n" : 0, "?" : np.NaN})

print(df.shape)
df.head()

(435, 17)


party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [18]:
# Make two dataframe subsets

df_rep = df[df["party"] == "republican"]
df_dem = df[df["party"] == "democrat"]

In [19]:
df_dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [20]:
df_rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

## Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level)

### Null Hypothesis:

Democrat support for the `handicapped infants` bill is not more than republican support for the `handicapped infants` bill

### Alternative Hypothesis:

Democrat support for the `handicapped infants` bill is more than republican support for the `handicapped infants` bill

### Confidence Level: 99%

In [22]:
stats.ttest_ind(df_dem["handicapped-infants"], df_rep["handicapped-infants"], nan_policy="omit")

Ttest_indResult(statistic=9.205264294809222, pvalue=1.613440327937243e-18)

### Conclusion

With a really significant t-score of 9.2 and a p-values of 1.6e-18, I reject the null hypothesis that democrat support for the `handicapped infants` bill is not more than republican support for the `handicapped infants` bill.

I suggest with 99% certainty that the democrats support the `handicapped infants` bill more than the republicans.

## Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level)

### Null Hypothesis:

Republican support for the `religious groups` bill is not more than democrat support for the `religious groups` bill

### Alternative Hypothesis:

Republican support for the `religious groups` bill is more than democrat support for the `religious groups` bill

### Confidence Level: 99%

In [25]:
stats.ttest_ind(df_rep["religious-groups"], df_dem["religious-groups"], nan_policy="omit")

Ttest_indResult(statistic=9.737575825219457, pvalue=2.3936722520597287e-20)

### Conclusion

With a really significant t-score of 9.74 and a p-values of 2.4e-20, I reject the null hypothesis that republican support for the `religious groups` bill is not more than democrat support for the `religious groups` bill.

I suggest with 99% certainty that the republicans support the `religious groups` bill more than the democrats.

## Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

### Null Hypothesis:

Republican support for the `water project` bill is the same as democrat support for the `water project` bill

### Alternative Hypothesis:

Republican support for the `water project` bill is not the same as democrat support for the `water project` bill

### Confidence Level: 90%

In [27]:
stats.ttest_ind(df_rep["water-project"], df_dem["water-project"], nan_policy="omit")

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

### Conclusion:

With a t-score of 0.089 and a p-value of 0.929, I fail to reject the null hypothesis that republican support for the `water project` bill is the same as democrat support for the `water project` bill with 90% certainty.

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [29]:
column_headers

['party',
 'handicapped-infants',
 'water-project',
 'budget',
 'physician-fee-freeze',
 'el-salvador-aid',
 'religious-groups',
 'anti-satellite-ban',
 'aid-to-contras',
 'mx-missile',
 'immigration',
 'synfuels',
 'education',
 'right-to-sue',
 'crime',
 'duty-free',
 'south-africa']

In [32]:
# Iterate t-scores for all policies across both parties
# Store in a list
t_scores = [stats.ttest_ind(df_dem[col], df_rep[col], nan_policy="omit")[0] for col in column_headers[1:]]

t_scores

[9.205264294809222,
 -0.08896538137868286,
 23.21277691701378,
 -49.36708157301406,
 -21.13669261173219,
 -9.737575825219457,
 12.526187929077842,
 18.052093200819733,
 16.437503268542994,
 -1.7359117329695164,
 8.293603989407588,
 -20.500685724563073,
 -13.51064251060933,
 -16.342085656197696,
 12.853146132542978,
 6.849454815841208]

In [33]:
# Iterate p-values for all policies across both parties
# Store in a list
p_values = [stats.ttest_ind(df_dem[col], df_rep[col], nan_policy="omit")[1] for col in column_headers[1:]]

p_values

[1.613440327937243e-18,
 0.9291556823993485,
 2.0703402795404463e-77,
 1.994262314074344e-177,
 5.600520111729011e-68,
 2.3936722520597287e-20,
 8.521033017443867e-31,
 2.82471841372357e-54,
 5.03079265310811e-47,
 0.08330248490425066,
 1.5759322301054064e-15,
 1.8834203990450192e-64,
 1.2278581709672758e-34,
 9.952342705606092e-47,
 5.997697174347365e-32,
 3.652674361672226e-11]

In [34]:
# Append both scores of the statistic to a nested list

scores = [list(i) for i in zip(t_scores, p_values)]

In [36]:
# Show statistic scores in a dataframe
df_scores = pd.DataFrame(scores, index=column_headers[1:], columns=["t-score", "p-value"])

df_scores

t-score        p-value
handicapped-infants    9.205264   1.613440e-18
water-project         -0.088965   9.291557e-01
budget                23.212777   2.070340e-77
physician-fee-freeze -49.367082  1.994262e-177
el-salvador-aid      -21.136693   5.600520e-68
religious-groups      -9.737576   2.393672e-20
anti-satellite-ban    12.526188   8.521033e-31
aid-to-contras        18.052093   2.824718e-54
mx-missile            16.437503   5.030793e-47
immigration           -1.735912   8.330248e-02
synfuels               8.293604   1.575932e-15
education            -20.500686   1.883420e-64
right-to-sue         -13.510643   1.227858e-34
crime                -16.342086   9.952343e-47
duty-free             12.853146   5.997697e-32
south-africa           6.849455   3.652674e-11

Note that each score is with respect to the Democratic party



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE